File: Medium/LC1026-Maximum-Difference-Between-Node-and-Ancestor/code.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        def dfs(root, maxi, mini):
            if not root:
                return maxi-mini
            maxi = max(maxi, root.val)
            mini = min(mini, root.val)
            return max(dfs(root.left, maxi, mini), dfs(root.right, maxi, mini))
        return maxDiff(root, root.val, root.val)

File: Medium/LC1026-Maximum-Difference-Between-Node-and-Ancestor/code.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    int maxDifference(TreeNode* root, int maxi, int mini) {
        if (root == NULL) {
            return (maxi - mini);
        }
        maxi = max(maxi, root->val);
        mini = min(mini, root->val);
        return max(maxDifference(root->left, maxi, mini), maxDifference(root->right, maxi, mini));
    }
    int maxAncestorDiff(TreeNode* root) {
        return maxDifference(root, root->val, root->val);
    }
};
